### DISCUSSION TITLE
Is there an integrated way to check if a promise was already resolved? Thoughts on new settleNow() helper

### DISCUSSION THREAD

**Discussion by marvinhinz** [reactphp/reactphp#571] [Q&A]
In my use case im inside a coroutine doing various async calls to control a vending machine during product vending (turning spirals, moving the lift, checking optical barrier etc.) and i start a promise that watches the result of the optical barrier check first (it will resolve only if the barrier was triggered, it retries api calls until the return value changes). 

Now after moving the lift i yield the opticalPromise (wrapped in a timeout). But what if i wanted to enforce the promise to fail right now, if its not resolved already. I couldnt find an integrated way to check if a promise was already resolved. If that was possible, an async helper function like "settleNow($promise)" could be very handy. Maybe im missing an obvious solution?

**Comment by SimonFrings** [comment]
Hey @marvinhinz, thanks for bringing this up :+1:

From what you're describing, it sounds like you want to check if a promise has settled in order to see if it's still possible to "make it fail." If I understand this correctly, then I think you're looking for the ability to [cancel the promise](https://github.com/reactphp/promise?tab=readme-ov-file#promiseinterfacecancel). This is also described in our documentation as follows:

> `$promise->cancel();`
>
> The `cancel()` method notifies the creator of the promise that there is no further interest in the results of the operation.
>
> Once a promise is settled (either fulfilled or rejected), calling `cancel()` on a promise has no effect.

As you can see, when using `cancel()`, we don't have to check if the promise is settled or not; it either cancels the still pending promise or does nothing instead.

I hope this answers your question :+1:

**Comment by marvinhinz** [comment]
Hey @SimonFrings thanks for your response, thats not quite what i had in mind, but i think it might achieve the same result :) 

In a nutshell i start an async task that might resolve or fail after timeout in the future, before yielding this promise to wait for it to finish i do other work. That work will most likely complete before the timeout of the async is reached, but at that point i dont really need to wait much longer. I guess if reactphp promises had a simple check function like $`promise->hasSettled()` or `settleNow($promise)` i could just check for that. 

Maybe wrapping it in another timeout `timeout($promise, 0)->finally();` might work the same 🤔

**Comment by WyriHaximus** [comment]
Please keep in mind that a promise is an intermedia object representing something for communication purposes. Not the action you are performing. If you want to fail/stop/interrupt that action you should do that either through rigging up `cancel` or calling it on the action you are performing directly, and the promise should resolve as a result.

Have you looked at the deferred object at https://reactphp.org/promise/#deferred-1 ? It grants you a different way of controlling the promise and gives you a direct way to reject it.

You can detect if a promise has been already resolve by doing this:

```php
$resolved = false;
$promise->then(static function () use (&$resolved): void {
    $resolved = true;
});

// $resolved will be true here if the promise has already been resolved
```
